File: app/agent_computer/src/core/agent-home-paths.ts
import { join, normalize, resolve } from 'node:path'
import { pathIsWithin } from './path-boundary'
import { modelIntegerIDFromWire } from './model-integer-id'

export const AGENTS_ROOT = '/agents'
export const WORKER_SHARE_ROOT = '/var/share'
// Codex SQLite state must live on Worker-local disk: WAL mode depends on
// same-host shared memory and is not usable on the shared network filesystem.
// The state is a rebuildable thread-recovery cache — losing it with the Worker
// falls into the existing "recreate a missing runtime chain" contract — while
// PostgreSQL keeps lifecycle truth and the shared filesystem keeps workspaces.
// Local state also makes per-(Agent, Worker) runtime shards structurally
// exclusive, which is what lets one Agent's Jobs run on several Workers.
// Read at call time so tests can point each fixture at its own root.
export function codexStateRoot(): string {
  return process.env.ANKOLE_CODEX_STATE_ROOT || '/var/lib/ankole/codex'
}
export const BUILTIN_SKILLS_ROOT = '/repo/app/library'
export const INTERNAL_SKILLS_ROOT = '/repo/internals/skills'
export const AGENT_UID_PATTERN = /^[a-z0-9][a-z0-9._-]{0,95}$/

export type AgentHomePaths = {
  agentKey: string
  home: string
  codexHome: string
  soul: string
  mission: string
  design: string
  userFiles: string
  installedSkills: string
  sessions: string
  jobs: string
}

export function assertAgentKey(value: string): string {
  if (!AGENT_UID_PATTERN.test(value)) {
    throw new Error('Agent UID must match ^[a-z0-9][a-z0-9._-]{0,95}$ before it can own files')
  }
  return value
}

export function agentHomePaths(agentsRoot: string, agentUID: string): AgentHomePaths {
  const agentKey = assertAgentKey(agentUID)
  const home = join(resolve(agentsRoot), agentKey)
  return {
    agentKey,
    home,
    codexHome: join(resolve(codexStateRoot()), agentKey, '.codex'),
    soul: join(home, 'SOUL.md'),
    mission: join(home, 'MISSION.md'),
    design: join(home, 'DESIGN.md'),
    userFiles: join(home, 'user-files'),
    installedSkills: join(home, 'installed-skills'),
    sessions: join(home, 'sessions'),
    jobs: join(home, 'jobs')
  }
}

export function sessionWorkspacePath(agentsRoot: string, agentUID: string, workspaceID: number): string {
  if (!Number.isSafeInteger(workspaceID) || workspaceID < 10_000) {
    throw new Error('Session workspace ID must be a model-safe integer starting at 10000')
  }
  return join(agentHomePaths(agentsRoot, agentUID).sessions, String(workspaceID))
}

export function jobWorkspacePath(agentsRoot: string, agentUID: string, jobID: string): string {
  modelIntegerIDFromWire(jobID, 'background Agent Job id')
  return join(agentHomePaths(agentsRoot, agentUID).jobs, jobID)
}

export function resolveAgentHomePath(
  agentHome: string,
  path: string,
  options: { cwd?: string; errorMessage?: string } = {}
): string {
  const home = resolve(agentHome)
  const errorMessage = options.errorMessage ?? 'path escapes Agent Home'
  const cwd = resolve(options.cwd ?? home)
  if (!pathIsWithin(home, cwd)) throw new Error(errorMessage)

  const expanded = path === '~' ? home : path.startsWith('~/') ? join(home, path.slice(2)) : path
  const resolved = expanded.startsWith('/') ? resolve(normalize(expanded)) : resolve(cwd, normalize(expanded))
  if (!pathIsWithin(home, resolved)) throw new Error(errorMessage)
  return resolved
}

export function insideAgentHome(agentHome: string, path: string): boolean {
  return pathIsWithin(resolve(agentHome), resolve(path))
}

export function assertPathInsideAgentHome(
  agentHome: string,
  path: string,
  message = 'path escapes Agent Home'
): string {
  const resolved = resolve(path)
  if (!insideAgentHome(agentHome, resolved)) throw new Error(message)
  return resolved
}

export interface SanitizePathSegmentOptions {
  fallback?: string
  replacement?: string
}

const safePathSegmentMaxLength = 96

export function sanitizePathSegment(value: string, options: SanitizePathSegmentOptions = {}): string {
  const replacement = options.replacement ?? '-'
  const fallback = options.fallback ?? 'default'
  const trimReplacement = replacement === '-'
  let safe = value.trim().replace(/[^a-zA-Z0-9._-]+/g, replacement)

  if (trimReplacement && replacement.length > 0) {
    const escaped = replacement.replace(/[.*+?^${}()|[\]\\]/g, '\\$&')
    safe = safe.replace(new RegExp(`^${escaped}+|${escaped}+$`, 'g'), '')
  }

  safe = safe.slice(0, safePathSegmentMaxLength)
  return safe || fallback
}

File: app/control_plane/lib/ankole/agent_home_paths.ex
defmodule Ankole.AgentHomePaths do
  @moduledoc """
  Canonical paths shared by the control plane and Agent Computer workers.

  Paths returned here are the real absolute paths visible to Codex. PostgreSQL
  remains authoritative for domain documents; files under Agent Home are runtime projections.
  """

  @agents_root "/agents"
  @agent_key ~r/\A[a-z0-9][a-z0-9._-]{0,95}\z/
  @documents %{"soul" => "SOUL.md", "mission" => "MISSION.md", "design" => "DESIGN.md"}

  @spec validate_agent_uid(term()) :: :ok | {:error, :invalid_agent_home_uid}
  def validate_agent_uid(uid) when is_binary(uid) do
    if Regex.match?(@agent_key, uid), do: :ok, else: {:error, :invalid_agent_home_uid}
  end

  def validate_agent_uid(_uid), do: {:error, :invalid_agent_home_uid}

  @spec home(String.t()) :: String.t()
  def home(agent_uid), do: join_agent(agent_uid, [])

  @spec user_files(String.t()) :: String.t()
  def user_files(agent_uid), do: join_agent(agent_uid, ["user-files"])

  @spec installed_skills(String.t()) :: String.t()
  def installed_skills(agent_uid), do: join_agent(agent_uid, ["installed-skills"])

  @spec session_workspace(String.t(), pos_integer()) :: String.t()
  def session_workspace(agent_uid, workspace_id),
    do: join_agent(agent_uid, ["sessions", workspace_key!(workspace_id)])

  @spec job_workspace(String.t(), pos_integer()) :: String.t()
  def job_workspace(agent_uid, job_id)
      when is_integer(job_id) and job_id in 1000..9_007_199_254_740_991,
      do: join_agent(agent_uid, ["jobs", Integer.to_string(job_id)])

  def job_workspace(_agent_uid, _job_id),
    do: raise(ArgumentError, "background Agent Job id must be a model-safe integer")

  @spec document(String.t(), String.t()) :: String.t()
  def document(agent_uid, kind), do: join_agent(agent_uid, [Map.fetch!(@documents, kind)])

  @spec user_files_lane_path(String.t(), String.t()) :: String.t()
  def user_files_lane_path(agent_uid, relative_path),
    do: lane_path(agent_uid, "user-files", relative_path)

  @spec installed_skills_lane_path(String.t(), String.t()) :: String.t()
  def installed_skills_lane_path(agent_uid, relative_path),
    do: lane_path(agent_uid, "installed-skills", relative_path)

  @spec session_lane_path(String.t(), pos_integer(), String.t()) :: String.t()
  def session_lane_path(agent_uid, workspace_id, relative_path),
    do: lane_path(agent_uid, "sessions/#{workspace_key!(workspace_id)}", relative_path)

  @spec document_lane_path(String.t(), String.t()) :: String.t()
  def document_lane_path(agent_uid, kind),
    do: Path.join(safe_agent_uid!(agent_uid), Map.fetch!(@documents, kind))

  defp join_agent(agent_uid, suffix),
    do: Path.join([@agents_root, safe_agent_uid!(agent_uid) | suffix])

  defp workspace_key!(workspace_id)
       when is_integer(workspace_id) and workspace_id in 10_000..9_007_199_254_740_991,
       do: Integer.to_string(workspace_id)

  defp workspace_key!(_workspace_id),
    do:
      raise(ArgumentError, "session workspace id must be a model-safe integer starting at 10000")

  defp lane_path(agent_uid, root, relative_path) do
    relative_path = String.trim_leading(relative_path, "/")
    Path.join([safe_agent_uid!(agent_uid), root, relative_path])
  end

  defp safe_agent_uid!(agent_uid) do
    case validate_agent_uid(agent_uid) do
      :ok -> agent_uid
      {:error, _reason} -> raise ArgumentError, "agent UID cannot own an Agent Home"
    end
  end
end
